jgrapht-core/src/main/java/org/jgrapht/alg/treedecomposition/IntervalgraphDecomposition.java

```diff
@@ -0,0 +1,347 @@
+        decomposition = new DefaultDirectedGraph<Integer, DefaultEdge>(DefaultEdge.class);
+        decompositionIntervalMap = new HashMap<Integer, Set<V>>();
+
+        // create root
+        currentVertex = 0;
+        root = 0;
+        currentSet = new HashSet<>();
+        decompositionIntervalMap.put(currentVertex, currentSet);
+        decomposition.addVertex(currentVertex);
+    }
+
+    /**
+     * Method for adding a new root as a joint node
+     */
+    private void addNewRoot()
+    {
+        Set<V> nextVertex = new HashSet<>();
+        currentSet = nextVertex;
+        
+        // new root
+        decomposition.addVertex(currentVertex + 1);
+        decompositionIntervalMap.put(currentVertex + 1, nextVertex);
+        
+        // set current Vertex
+        currentVertex++;
+        
+        // new current root
+        decomposition.addVertex(currentVertex + 1);
+        decompositionIntervalMap.put(currentVertex + 1, nextVertex);
+        
+        // make joint edges
+        decomposition.addEdge(currentVertex, currentVertex+1);
+        decomposition.addEdge(currentVertex, root);
+        root = currentVertex;
+
+        // set current Vertex
+        currentVertex++;
+    }
+
+    /**
+     * Method for adding introducing nodes
+     * 
+     * @param vertex the vertex, which is introduced
+     */
+    private void addIntroduce(Interval<T> interval)
+    {
+        Set<V> nextVertex = new HashSet<>(currentSet);
+        nextVertex.add(intervalToVertexMap.get(interval));
+        currentSet = nextVertex;
+        decomposition.addVertex(currentVertex + 1);
+        decomposition.addEdge(currentVertex, currentVertex + 1);
+        decompositionIntervalMap.put(currentVertex + 1, nextVertex);
+
+        // set current Vertex
+        currentVertex++;
+    }
+
+    /**
+     * method for adding forget nodes
+     * 
+     * @param vertex the vertex, which is introduced
+     */
+    private void addForget(Interval<T> interval)
+    {
+        Set<V> nextVertex = new HashSet<>(currentSet);
+        nextVertex.remove(intervalToVertexMap.get(interval));
+        currentSet = nextVertex;
+        decomposition.addVertex(currentVertex + 1);
+        decomposition.addEdge(currentVertex, currentVertex + 1);
+        decompositionIntervalMap.put(currentVertex + 1, nextVertex);
+
+        // set current Vertex
+        currentVertex++;
+    }
+
+    /**
+     * getter for the decomposition as an directed graph
+     * 
+     * @return the computed decomposition
+     */
+    public Graph<Integer, DefaultEdge> getDecomposition()
+    {
+        if (decomposition == null)
+            computeNiceDecomposition();
+        return decomposition;
+    }
+
+    /**
+     * getter for the map from integer nodes of the decomposition to the intervals of the interval
+     * graph
+     * 
+     * @return a nodes to interval map
+     */
+    public Map<Integer, Set<V>> getMap()
+    {
+        if (decompositionIntervalMap == null)
```

Comment:
actually I think this is not elegant enough. The only responsibility of this class is to compute the decomposition, which can already be done when calling the constructor from outside. After the constructor returns, the result should be already there, therefore calling get methods should be able to get the result instead of doing null-check in each method.